src/rgw/services/svc_rados.cc

```diff
@@ -1,306 +1,410 @@
+    ec = w.wait();
+  }
+  if (ec) {
+    if (ec == boost::system::errc::result_out_of_range) {
+      ldout(cct, 0)
+        << __func__
+        << " ERROR: RADOS::RADOS::create_pool returned " << ec
+        << " (this can be due to a pool or placement group misconfiguration, e.g."
+        << " pg_num < pgp_num or mon_max_pg_per_osd exceeded)"
+        << dendl;
     }
-    retcodes->push_back(ret);
+    return ec;
   }
-  if (error) {
-    return 0;
+  {
+    ceph::waiter<boost::system::error_code> w;
+    rados.enable_application(p.name, pg_pool_t::APPLICATION_NAME_RGW, false,
+                             w.ref());
+    ec = w.wait();
   }
-
-  completions.clear();
-  for (auto &io_ctx : io_ctxs) {
-    librados::PoolAsyncCompletion *c =
-      librados::Rados::pool_async_create_completion();
-    completions.push_back(c);
-    int ret = io_ctx.application_enable_async(pg_pool_t::APPLICATION_NAME_RGW,
-                                              false, c);
-    ceph_assert(ret == 0);
+  if (ec && ec != boost::system::errc::operation_not_supported) {
+    return ec;
   }
 
-  retcodes->clear();
-  for (auto c : completions) {
-    c->wait();
-    int ret = c->get_return_value();
-    if (ret == -EOPNOTSUPP) {
-      ret = 0;
-    } else if (ret < 0) {
-      ldout(rados_svc->cct, 0) << "WARNING: async application_enable returned " << ret
-                    << dendl;
-      error = true;
-    }
-    c->release();
-    retcodes->push_back(ret);
-  }
-  return 0;
+  return {};
 }
 
-int RGWSI_RADOS::Pool::lookup()
-{
-  librados::Rados *rad = rados_svc->get_rados_handle();
-  int ret = rad->pool_lookup(pool.name.c_str());
-  if (ret < 0) {
-    return ret;
+boost::system::error_code
+RGWSI_RADOS::create_pools(const std::vector<rgw_pool>& pools,
+                          std::vector<boost::system::error_code>* e,
+                          optional_yield y) {
+  // TODO: Figure out how to queue up a whole bunch of operations on a
+  // single coroutine rather than serializing them.
+  boost::system::error_code tec;
+
+  if (e)
+    e->clear();
+  for (const auto& p : pools) {
+    auto ec = create_pool(p, y);
+    if (ec)
+      tec = ec;
+    if (e)
+      e->push_back(ec);
   }
+  return tec;
+}
 
-  return 0;
+RGWSI_RADOS::Pool::List RGWSI_RADOS::Pool::list(RGWAccessListFilter filter) {
+  return List(*this, std::move(filter));
 }
 
-int RGWSI_RADOS::Pool::List::init(const string& marker, RGWAccessListFilter *filter)
+boost::system::error_code
+RGWSI_RADOS::Pool::List::get_next(int max,
+                                  std::vector<std::string>* oids,
+                                  bool* is_truncated,
+                                  optional_yield y)
 {
-  if (ctx.initialized) {
-    return -EINVAL;
-  }
-
-  int r = pool.rados_svc->open_pool_ctx(pool.pool, ctx.ioctx);
-  if (r < 0) {
-    return r;
+  if (iter == RADOS::EnumerationCursor::begin())
+    return ceph::to_error_code(ENOENT);
```

Comment:
iter starts at `begin()`, so this always fails. not sure what this is trying to do, but remove the check seems to work

Reply:
It was supposed to be ::end. I think this was an example of bug-by-sleep-deprivation.